File: problems-extend/0101-0200/0116_populating-next-right-pointers-in-each-node/solution.java
class Solution {

    public NodeWithNext connect(NodeWithNext root) {
        if (root == null) return null;
        NodeWithNext level = root;
        while (level.left != null) {
            NodeWithNext head = level;
            while (head != null) {
                head.left.next = head.right;
                if (head.next != null) head.right.next = head.next.left;
                head = head.next;
            }
            level = level.left;
        }
        return root;
    }
}

File: problems-extend/0101-0200/0116_populating-next-right-pointers-in-each-node/solution.go
func connect(root *NodeWithNext) *NodeWithNext {
	if root == nil {
		return nil
	}
	for level := root; level.Left != nil; level = level.Left {
		for head := level; head != nil; head = head.Next {
			head.Left.Next = head.Right
			if head.Next != nil {
				head.Right.Next = head.Next.Left
			}
		}
	}
	return root
}
